File: src/RpgmDecrypt.Core/Log.fs
namespace RpgmDecrypt.Core

/// Logging sinks.
///
/// Two formats:
///   * `Human` — ANSI-friendly progress lines + summary.
///   * `Json`  — newline-delimited JSON for piping into `jq`.
module Log =

    type Format = Human | Json

    type Event =
        | Walked      of path: string * size: int64
        | Detected    of path: string * format: string
        | KeyFound    of source: string
        | Decrypt     of inputPath: string * outputPath: string * format: string
        | PassThrough of path: string
        | Skipped     of path: string * reason: string
        | Failed      of path: string * reason: string
        | Summary     of RunSummary

    let private escape (s: string) : string =
        let sb = System.Text.StringBuilder(s.Length + 8)
        for c in s do
            match c with
            | '\\' -> sb.Append("\\\\") |> ignore
            | '"'  -> sb.Append("\\\"") |> ignore
            | '\n' -> sb.Append("\\n") |> ignore
            | '\r' -> sb.Append("\\r") |> ignore
            | '\t' -> sb.Append("\\t") |> ignore
            | c when int c < 32 -> sb.Append(' ') |> ignore
            | c    -> sb.Append(c) |> ignore
        sb.ToString()

    let private hrFmt =
        function
        | XP -> "XP" | VX -> "VX" | VXAce -> "VXAce" | MV -> "MV" | MZ -> "MZ"

    let private summaryToJson (s: RunSummary) : string =
        let perFmt =
            s.PerFormat
            |> Map.toList
            |> List.map (fun (k, v) -> sprintf "\"%s\":%d" (hrFmt k) v)
            |> String.concat ","
        sprintf
            """{"kind":"summary","started":"%s","finished":"%s","scanned":%d,"decrypted":%d,"passthrough":%d,"skipped":%d,"failed":%d,"key_source":"%s","per_format":{%s}}"""
            (escape (s.StartedAt.ToString("o")))
            (escape (s.FinishedAt.ToString("o")))
            s.InputsScanned
            s.DecryptedCount
            s.PassedThroughCount
            s.SkippedCount
            s.FailedCount
            (escape s.KeySource)
            perFmt

    let private eventToJson (e: Event) : string =
        match e with
        | Walked(p, sz) -> sprintf """{"kind":"walked","path":"%s","size":%d}""" (escape p) sz
        | Detected(p, fmt_) -> sprintf """{"kind":"detected","path":"%s","format":"%s"}""" (escape p) (escape fmt_)
        | KeyFound src     -> sprintf """{"kind":"key_found","source":"%s"}""" (escape src)
        | Decrypt(i, o, fmt_) -> sprintf """{"kind":"decrypt","input":"%s","output":"%s","format":"%s"}""" (escape i) (escape o) (escape fmt_)
        | PassThrough p    -> sprintf """{"kind":"passthrough","path":"%s"}""" (escape p)
        | Skipped(p, r)    -> sprintf """{"kind":"skipped","path":"%s","reason":"%s"}""" (escape p) (escape r)
        | Failed(p, r)     -> sprintf """{"kind":"failed","path":"%s","reason":"%s"}""" (escape p) (escape r)
        | Summary s        -> summaryToJson s

    let private humanSummary (s: RunSummary) : string =
        let dur = s.FinishedAt - s.StartedAt
        let perFmtLine =
            s.PerFormat
            |> Map.toList
            |> List.map (fun (k, v) -> sprintf "%s=%d" (hrFmt k) v)
            |> String.concat " "
        sprintf
            "scanned: %d\ndecrypted: %d\npass-through: %d\nskipped: %d\nfailed: %d\nkey source: %s\nduration: %.2fs\nby format: %s"
            s.InputsScanned
            s.DecryptedCount
            s.PassedThroughCount
            s.SkippedCount
            s.FailedCount
            s.KeySource
            dur.TotalSeconds
            perFmtLine

    /// Emit one event to STDERR using chosen format.
    let emit (fmt: Format) (e: Event) : unit =
        match fmt with
        | Human ->
            match e with
            | Walked(p, sz)        -> eprintfn "walked %s (%d B)" p sz
            | Detected(p, fmt_)    -> eprintfn "  + detected %s as %s" p fmt_
            | KeyFound src         -> eprintfn "[key] %s" src
            | Decrypt(i, o, fmt_)  -> eprintfn "  > %s -> %s [%s]" i o fmt_
            | PassThrough p        -> eprintfn "  = %s (already plaintext)" p
            | Skipped(p, r)        -> eprintfn "  - skipped %s (%s)" p r
            | Failed(p, r)         -> eprintfn "  x failed %s (%s)" p r
            | Summary s            -> eprintfn "\n=== summary ===\n%s" (humanSummary s)
        | Json  -> eprintfn "%s" (eventToJson e)

File: ocaml/lib/log.ml
(* Source — Human + NDJSON sinks to stderr. *)

type format = Human | Json

type event =
  | Walked of string * int64
  | Detected of string * string
  | KeyFound of string
  | Decrypt of string * string * string
  | PassThrough of string
  | Skipped of string * string
  | Failed of string * string
  | Summary of Types.run_summary

let escape (s : string) : string =
  let b = Buffer.create (String.length s + 8) in
  String.iter
    (fun c ->
      match c with
      | '\\' -> Buffer.add_string b "\\\\"
      | '"' -> Buffer.add_string b "\\\""
      | '\n' -> Buffer.add_string b "\\n"
      | '\r' -> Buffer.add_string b "\\r"
      | '\t' -> Buffer.add_string b "\\t"
      | c when Char.code c < 32 -> Buffer.add_char b ' '
      | c -> Buffer.add_char b c)
    s;
  Buffer.contents b

let iso_of_time (t : float) : string =
  let tm = Unix.gmtime t in
  Printf.sprintf "%04d-%02d-%02dT%02d:%02d:%02dZ"
    (tm.Unix.tm_year + 1900) (tm.Unix.tm_mon + 1) tm.Unix.tm_mday tm.Unix.tm_hour
    tm.Unix.tm_min tm.Unix.tm_sec

let per_fmt_json (pf : (Types.format * int) list) : string =
  pf
  |> List.map (fun (k, v) ->
         Printf.sprintf "\"%s\":%d" (Types.format_to_string k) v)
  |> String.concat ","

let summary_to_json (s : Types.run_summary) : string =
  Printf.sprintf
    "{\"kind\":\"summary\",\"started\":\"%s\",\"finished\":\"%s\",\"scanned\":%d,\"decrypted\":%d,\"passthrough\":%d,\"skipped\":%d,\"failed\":%d,\"key_source\":\"%s\",\"per_format\":{%s}}"
    (escape (iso_of_time s.Types.started_at))
    (escape (iso_of_time s.Types.finished_at))
    s.Types.inputs_scanned s.Types.decrypted_count s.Types.passed_through_count
    s.Types.skipped_count s.Types.failed_count (escape s.Types.key_source)
    (per_fmt_json s.Types.per_format)

let event_to_json (e : event) : string =
  match e with
  | Walked (p, sz) ->
      Printf.sprintf "{\"kind\":\"walked\",\"path\":\"%s\",\"size\":%Ld}"
        (escape p) sz
  | Detected (p, fmt) ->
      Printf.sprintf "{\"kind\":\"detected\",\"path\":\"%s\",\"format\":\"%s\"}"
        (escape p) (escape fmt)
  | KeyFound src ->
      Printf.sprintf "{\"kind\":\"key_found\",\"source\":\"%s\"}" (escape src)
  | Decrypt (i, o, fmt) ->
      Printf.sprintf
        "{\"kind\":\"decrypt\",\"input\":\"%s\",\"output\":\"%s\",\"format\":\"%s\"}"
        (escape i) (escape o) (escape fmt)
  | PassThrough p ->
      Printf.sprintf "{\"kind\":\"passthrough\",\"path\":\"%s\"}" (escape p)
  | Skipped (p, r) ->
      Printf.sprintf "{\"kind\":\"skipped\",\"path\":\"%s\",\"reason\":\"%s\"}"
        (escape p) (escape r)
  | Failed (p, r) ->
      Printf.sprintf "{\"kind\":\"failed\",\"path\":\"%s\",\"reason\":\"%s\"}"
        (escape p) (escape r)
  | Summary s -> summary_to_json s

let human_summary (s : Types.run_summary) : string =
  let dur = s.Types.finished_at -. s.Types.started_at in
  let per_fmt_line =
    s.Types.per_format
    |> List.map (fun (k, v) ->
           Printf.sprintf "%s=%d" (Types.format_to_string k) v)
    |> String.concat " "
  in
  Printf.sprintf
    "scanned: %d\ndecrypted: %d\npass-through: %d\nskipped: %d\nfailed: %d\nkey \
     source: %s\nduration: %.2fs\nby format: %s"
    s.Types.inputs_scanned s.Types.decrypted_count s.Types.passed_through_count
    s.Types.skipped_count s.Types.failed_count s.Types.key_source dur
    per_fmt_line

(** Emit one event to STDERR using the chosen format. *)
let emit (fmt : format) (e : event) : unit =
  match fmt with
  | Human -> (
      match e with
      | Walked (p, sz) -> Printf.eprintf "walked %s (%Ld B)\n" p sz
      | Detected (p, f) -> Printf.eprintf "  + detected %s as %s\n" p f
      | KeyFound src -> Printf.eprintf "[key] %s\n" src
      | Decrypt (i, o, f) -> Printf.eprintf "  > %s -> %s [%s]\n" i o f
      | PassThrough p -> Printf.eprintf "  = %s (already plaintext)\n" p
      | Skipped (p, r) -> Printf.eprintf "  - skipped %s (%s)\n" p r
      | Failed (p, r) -> Printf.eprintf "  x failed %s (%s)\n" p r
      | Summary s -> Printf.eprintf "\n=== summary ===\n%s\n" (human_summary s))
  | Json -> Printf.eprintf "%s\n" (event_to_json e)
